Gems/SceneProcessing/Code/Source/Generation/Components/MeshOptimizer/MeshOptimizerComponent.cpp

```diff
@@ -96,6 +96,85 @@ namespace AZ::SceneGenerationComponents
+    // constant time lookup to map from an unwelded vertex index to the welded one.
+    // The welding works by rounding the vertex's position to the given position tolerance, then uses that rounded
+    // Vector3 as a key into a unordered_map.
+    template <class MeshDataType>
+    class Vector3Map
+        : private AZStd::unordered_map<AZ::Vector3, AZ::u32>
+    {
+    public:
+        Vector3Map(const MeshDataType* meshData, bool hasBlendShapes, float positionTolerance)
+            : m_meshData(meshData)
+            , m_hasBlendShapes(hasBlendShapes)
+            , m_positionTolerance(positionTolerance)
+            , m_positionToleranceReciprocal(1.0f / positionTolerance)
+        {
+        }
+
+        using AZStd::unordered_map<AZ::Vector3, AZ::u32>::reserve;
+
+        AZ::u32 operator[](const AZ::u32 vertexIndex)
+        {
+            if (m_hasBlendShapes)
+            {
+                // Don't attempt to weld similar vertices if there's blendshapes
+                // Welding the vertices here based on position could cause the vertices of a base shape to be welded,
+                // and the vertices of the blendshape to not be welded, resulting in a vertex count mismatch between
+                // the two
+                return m_meshData->GetUsedPointIndexForControlPoint(m_meshData->GetControlPointIndex(vertexIndex));
+            }
+
+            const auto& [iter, didInsert] = try_emplace(GetPositionForIndex(vertexIndex), m_currentOriginalVertexIndex);
+            if (didInsert)
+            {
+                ++m_currentOriginalVertexIndex;
+            }
+            return iter->second;
+        }
+
+        [[nodiscard]] AZ::u32 at(const AZ::u32 vertexIndex) const
+        {
+            if (m_hasBlendShapes)
+            {
+                // Don't attempt to weld similar vertices if there's blendshapes
+                // Welding the vertices here based on position could cause the vertices of a base shape to be welded,
+                // and the vertices of the blendshape to not be welded, resulting in a vertex count mismatch between
+                // the two
+                return m_meshData->GetUsedPointIndexForControlPoint(m_meshData->GetControlPointIndex(vertexIndex));
+            }
+
+            auto iter = find(GetPositionForIndex(vertexIndex));
+            AZSTD_CONTAINER_ASSERT(iter != end(), "Element with key is not present");
+            return iter->second;
+        }
+
+    private:
+
+        AZ::Vector3 GetPositionForIndex(const AZ::u32 vertexIndex) const
+        {
+            // Round the vertex position so that a float comparison can be made with entires in the map
+            // pos = floor( x * 10 + 0.5) * 0.1
+            return AZ::Vector3(
+                AZ::Simd::Vec3::Floor(
+                    (m_meshData->GetPosition(vertexIndex) * m_positionToleranceReciprocal + AZ::Vector3(0.5f)).GetSimdValue()
+                )
+            ) * m_positionTolerance;
```

Comment:
This is great but also the risky part: I think we should get a new rule up for mesh optimizer settings, having the tolerance used to weld vertices in there. Even if we later might switch to AssImp's, we can reuse the rule then as we need some sort of control. Using a rule we also have the option to disable the mesh optimizer, in case there are issues with some asset. Then we can just let the user know to remove the rule.

We'd probably have to merge in the skinning rule into that one or reuse the skinning one and just rebranch it (that might actually be the better and easier way, taking backward compatibility into account).

Reply:
Whether we do a comparison on a rounded Vector3, or use a direct floating point comparison, the result will be the same: we don't know at runtime if the user intended the two vertices to be separated or not.